File: blockball-api/src/main/java/com/github/shynixn/blockball/api/business/entity/Ball.java
package com.github.shynixn.blockball.api.business.entity;

import com.github.shynixn.blockball.api.persistence.entity.BallMeta;

public interface Ball {

    /**
     * Kicks the ball with the given strength parameters.
     *
     * @param entity             entity
     * @param horizontalStrength horizontalStrength
     * @param verticalStrength   verticalStrength
     */
    void kick(Object entity, double horizontalStrength, double verticalStrength);

    /**
     * Kicks the ball with the default strength values.
     *
     * @param entity entity
     */
    void kick(Object entity);

    /**
     * Passes the ball with the given strength parameters.
     *
     * @param entity             entity
     * @param horizontalStrength horizontalStrength
     * @param verticalStrength   verticalStrength
     */
    void pass(Object entity, double horizontalStrength, double verticalStrength);

    /**
     * Passes the ball with the default strength values.
     *
     * @param entity entity
     */
    void pass(Object entity);

    /**
     * Shoots the ball in the during vector direction.
     *
     * @param vector vector
     */
    void setVelocity(Object vector);

    /**
     * Returns the velocity the ball is flying after a kick/pass or velocity manipulation.
     * @return velocity
     */
    Object getVelocity();

    /**
     * Respawns the entity at the given location.
     *
     * @param location location
     */
    void spawn(Object location);

    /**
     * Damages the ball entity with the given amount.
     *
     * @param amount amount
     */
    void damage(double amount);

    /**
     * Removes the entity.
     */
    void remove();

    /**
     * Teleports the ball to the given location.
     *
     * @param location location
     */
    void teleport(Object location);

    /**
     * Returns the location of the ball.
     *
     * @return location
     */
    Object getLocation();

    /**
     * Returns if the ball entity is Dead.
     *
     * @return isDead
     */
    boolean isDead();

    /**
     * Returns the entity who is responsible for the ball design and skin.
     *
     * @return entity
     */
    Object getDesignEntity();

    /**
     * Returns the entity who is responsible for the hitbox.
     *
     * @return entity
     */
    Object getHitboxEntity();

    /**
     * Returns the metaData of the entity.
     *
     * @return meta
     */
    BallMeta getMeta();
}

File: src/main/java/com/github/shynixn/blockball/contract/Ball.kt
package com.github.shynixn.blockball.contract

import com.github.shynixn.blockball.entity.BallMeta
import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.util.EulerAngle
import org.bukkit.util.Vector

/**
 * Ball interface of the complex ball entity.
 */
interface Ball {
    /**
     * Gets the meta data.
     */
    val meta: BallMeta

    /**
     * Is the entity dead?
     */
    val isDead: Boolean

    /**
     * Entity id of the hitbox.
     */
    val hitBoxEntityId: Int

    /**
     * Entity id of the design.
     */
    val designEntityId: Int

    /**
     * Gets if the ball is on ground.
     */
    val isOnGround: Boolean

    /**
     * Teleports the ball to the given [location].
     */
    fun teleport(location: Location)

    /**
     * Gets the location of the ball.
     */
    fun getLocation(): Location

    /**
     * Gets the velocity of the ball.
     */
    fun  getVelocity(): Vector

    /**
     * Rotation of the visible ball in euler angles.
     */
    fun getRotation(): Vector

    /**
     * Shoot the ball by the given player.
     * The calculated velocity can be manipulated by the BallKickEvent.
     *
     * @param player
     */
    fun kickByPlayer(player: Player)

    /**
     * Pass the ball by the given player.
     * The calculated velocity can be manipulated by the BallKickEvent
     *
     * @param player
     */
    fun passByPlayer(player: Player)

    /**
     * Removes the ball.
     */
    fun remove()

    /**
     * Runnable. Should not be called directly.
     */
    fun run()
}
